File: apps/target_site/app/antibot.py
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum


class AntibotAction(str, Enum):
    ALLOW = "allow"
    DELAY = "delay"
    CHALLENGE = "challenge"
    FORBID = "forbid"
    RATE_LIMIT = "rate_limit"


@dataclass
class Visit:
    path: str
    created_at: datetime


@dataclass
class SessionState:
    session_id: str
    proxy_id: str
    visits: list[Visit] = field(default_factory=list)
    captcha_errors: int = 0
    saw_listing: bool = False


@dataclass(frozen=True)
class AntibotDecision:
    risk_score: int
    action: AntibotAction
    reason: str


class AntibotSimulator:
    def __init__(self) -> None:
        self.sessions: dict[str, SessionState] = {}

    def get_state(self, session_id: str, proxy_id: str) -> SessionState:
        key = f"{proxy_id}:{session_id}"
        if key not in self.sessions:
            self.sessions[key] = SessionState(session_id=session_id, proxy_id=proxy_id)
        return self.sessions[key]

    def record_captcha_error(self, session_id: str, proxy_id: str) -> None:
        state = self.get_state(session_id, proxy_id)
        state.captcha_errors += 1

    def evaluate(
        self,
        *,
        session_id: str,
        proxy_id: str,
        path: str,
        user_agent: str | None,
        has_clearance_cookie: bool,
        now: datetime | None = None,
    ) -> AntibotDecision:
        current_time = now or datetime.utcnow()
        state = self.get_state(session_id, proxy_id)
        state.visits = [
            visit for visit in state.visits if visit.created_at > current_time - timedelta(minutes=1)
        ]
        state.visits.append(Visit(path=path, created_at=current_time))

        if path.startswith("/items"):
            state.saw_listing = True

        score = 0
        reasons: list[str] = []

        if len(state.visits) >= 12:
            return AntibotDecision(95, AntibotAction.RATE_LIMIT, "muitas requisicoes em 1 minuto")

        if len(state.visits) >= 7:
            score += 35
            reasons.append("volume alto por minuto")

        if not has_clearance_cookie:
            score += 20
            reasons.append("sem cookie de sessao liberada")

        if not user_agent or "bot" in user_agent.lower():
            score += 20
            reasons.append("user-agent ausente ou suspeito")

        if path.startswith("/items/") and not state.saw_listing:
            score += 25
            reasons.append("acesso direto ao detalhe")

        if state.captcha_errors:
            score += min(30, state.captcha_errors * 15)
            reasons.append("erros de captcha anteriores")

        score = min(score, 100)
        reason = ", ".join(reasons) or "baixo risco"

        if score < 40:
            return AntibotDecision(score, AntibotAction.ALLOW, reason)
        if score < 70:
            return AntibotDecision(score, AntibotAction.DELAY, reason)
        if score < 90:
            return AntibotDecision(score, AntibotAction.CHALLENGE, reason)
        return AntibotDecision(score, AntibotAction.FORBID, reason)

File: apps/target_site/src/lib/antibot.ts
import {
  ANTI_BOT_HIGH_VOLUME_VISITS,
  ANTI_BOT_RATE_LIMIT_VISITS,
  ANTI_BOT_VISIT_WINDOW_MS
} from "./site-contracts";

export enum AntibotAction {
  Allow = "allow",
  Delay = "delay",
  Challenge = "challenge",
  Forbid = "forbid",
  RateLimit = "rate_limit"
}

export type AntibotDecision = {
  riskScore: number;
  action: AntibotAction;
  reason: string;
};

type Visit = {
  path: string;
  createdAt: Date;
};

type SessionState = {
  sessionId: string;
  proxyId: string;
  visits: Visit[];
  captchaErrors: number;
  sawListing: boolean;
};

type EvaluateInput = {
  sessionId: string;
  proxyId: string;
  path: string;
  userAgent: string | null;
  hasClearanceCookie: boolean;
  now?: Date;
};

export class AntibotSimulator {
  private readonly sessions = new Map<string, SessionState>();

  recordCaptchaError(sessionId: string, proxyId: string): void {
    this.getState(sessionId, proxyId).captchaErrors += 1;
  }

  evaluate(input: EvaluateInput): AntibotDecision {
    const currentTime = input.now ?? new Date();
    const state = this.getState(input.sessionId, input.proxyId);
    const oneMinuteAgo = currentTime.getTime() - ANTI_BOT_VISIT_WINDOW_MS;
    state.visits = state.visits.filter((visit) => visit.createdAt.getTime() > oneMinuteAgo);
    state.visits.push({ path: input.path, createdAt: currentTime });

    if (input.path.startsWith("/items")) {
      state.sawListing = true;
    }

    if (state.visits.length >= ANTI_BOT_RATE_LIMIT_VISITS) {
      return {
        riskScore: 95,
        action: AntibotAction.RateLimit,
        reason: "muitas requisicoes em 1 minuto"
      };
    }

    const reasons: string[] = [];
    let score = 0;

    if (state.visits.length >= ANTI_BOT_HIGH_VOLUME_VISITS) {
      score += 35;
      reasons.push("volume alto por minuto");
    }

    if (!input.hasClearanceCookie) {
      score += 20;
      reasons.push("sem cookie de sessao liberada");
    }

    if (!input.userAgent || input.userAgent.toLowerCase().includes("bot")) {
      score += 20;
      reasons.push("user-agent ausente ou suspeito");
    }

    if (input.path.startsWith("/items/") && !state.sawListing) {
      score += 25;
      reasons.push("acesso direto ao detalhe");
    }

    if (state.captchaErrors > 0) {
      score += Math.min(30, state.captchaErrors * 15);
      reasons.push("erros de captcha anteriores");
    }

    const riskScore = Math.min(score, 100);
    const reason = reasons.join(", ") || "baixo risco";

    if (riskScore < 40) {
      return { riskScore, action: AntibotAction.Allow, reason };
    }
    if (riskScore < 70) {
      return { riskScore, action: AntibotAction.Delay, reason };
    }
    if (riskScore < 90) {
      return { riskScore, action: AntibotAction.Challenge, reason };
    }
    return { riskScore, action: AntibotAction.Forbid, reason };
  }

  snapshot(): Array<SessionState & { visits: Visit[] }> {
    return Array.from(this.sessions.values()).map((state) => ({
      ...state,
      visits: [...state.visits]
    }));
  }

  private getState(sessionId: string, proxyId: string): SessionState {
    const key = `${proxyId}:${sessionId}`;
    const existing = this.sessions.get(key);
    if (existing) {
      return existing;
    }

    const state: SessionState = {
      sessionId,
      proxyId,
      visits: [],
      captchaErrors: 0,
      sawListing: false
    };
    this.sessions.set(key, state);
    return state;
  }
}

export const antibotSimulator = new AntibotSimulator();
